kernel/bpf/btf.c

```diff
@@ -5458,9 +5458,9 @@ static int btf_check_func_arg_match(struct bpf_verifier_env *env,
 			if (reg->type == PTR_TO_BTF_ID) {
 				reg_btf = reg->btf;
 				reg_ref_id = reg->btf_id;
-			} else if (reg2btf_ids[reg->type]) {
+			} else if (reg2btf_ids[base_type(reg->type)]) {
```

Comment:
There's an upstream diff here that should be called out I think.

Reply:
Good call.  I've added a note from the stable backport of this commit justifying the difference.

Reply:
This is missing apart of the conditional
``` (reg2btf_ids[base_type(reg->type)] && !type_flag(reg->type))))```
which is in the `553.16.1` code base
https://github.com/ctrliq/kernel-src-tree/blame/resf_kernel-4.18.0-553.16.1.el8_10/kernel/bpf/btf.c#L5461

Reply:
I didn't add it because of how this change was backported in stable

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=8c39925e98d498b9531343066ef82ae39e41adae

It is odd that it doesn't match 533.16.1 though.  I can change it to match if we want.  FWIW, in the upcoming PR for VULN-72 and VULN-144 I am going to backport 3363bd0cfbb80, and as such will need to backport the additional conditionals.

Reply:
Sorry for the delay I have a thousand tabs and several terminals up.

We'll go with the solution you found ... I'm tyring to find why redhat would have included the `!type_flag(reg->type)` or why the original authors did too

include/linux/bpf_verifier.h
```
    3 /* extract flags from an extended type. See bpf_type_flag in bpf.h. */
    2 static inline u32 type_flag(u32 type)
    1 {
  531         return type & ~BPF_BASE_TYPE_MASK;
    1 }
```

But based off the conversation in the first thread there wasn't much more follow up here
https://lore.kernel.org/bpf/CAADnVQJ-wMjyBQUYELYCjDTST8M5+TKRw2fi7nfrv79319fwog@mail.gmail.com/
```
> > >  /* Returns true if struct is composed of scalars, 4 levels of nesting allowed */
> > >  static bool __btf_type_is_scalar_struct(struct bpf_verifier_log *log,
> > > @@ -5688,7 +5696,7 @@ static int btf_check_func_arg_match(struct bpf_verifier_env *env,
> > >                         }
> > >                         if (check_ptr_off_reg(env, reg, regno))
> > >                                 return -EINVAL;
> > > -               } else if (is_kfunc && (reg->type == PTR_TO_BTF_ID || reg2btf_ids[reg->type])) {
> > > +               } else if (is_kfunc && (reg->type == PTR_TO_BTF_ID || reg2btf_ids(reg->type))) {
> >
> > Just use reg2btf_ids[base_type(reg->type)] instead?
>
> That would be incorrect I think, then we'd allow e.g. PTR_TO_TCP_SOCK_OR_NULL
> and treat it as PTR_TO_TCP_SOCK, while current code only wants to permit
> non-NULL variants for these three.

add && !type_flag(reg->type) ?
```

I can't come up with enough of a reason to block this so I'll consider it done.

> o backport 3363bd0

This will actually refactor a lot of this ... I also haven't found here they remove the ` !type_flag(reg->type)` either.

I'll resolve this and leave it as is.